pallets/fusion/src/lib.rs

```diff
@@ -0,0 +1,3044 @@
+		/// The overarching event type.
+		type RuntimeEvent: From<Event<Self>> + IsType<<Self as frame_system::Config>::RuntimeEvent>;
+
+		/// A sudo-able call.
+		type RuntimeCall: Parameter
+			+ UnfilteredDispatchable<RuntimeOrigin = Self::RuntimeOrigin>
+			+ GetDispatchInfo;
+
+		/// Type representing the weight of this pallet
+		type WeightInfo: WeightInfo;
+
+		/// Someone who can call the admin extrinsics.
+		type ApprovedOrigin: EnsureOrigin<Self::RuntimeOrigin>;
+
+		/// Currency type for this pallet.
+		type Currency: LockableCurrency<Self::AccountId, Moment = BlockNumberFor<Self>>;
+
+		/// Convert a balance into a number used for election calculation.
+		type CurrencyToVote: sp_staking::currency_to_vote::CurrencyToVote<BalanceOf<Self>>;
+
+		/// The destination for rewards that were not claimed after 'HistoryDepth' eras.
+		type RewardRemainder: OnUnbalanced<NegativeImbalanceOf<Self>>;
+
+		/// Pallet id used to derive accounts used by the pallet
+		#[pallet::constant]
+		type PalletId: Get<PalletId>;
+
+		/// Maximum allowed for the currency name
+		#[pallet::constant]
+		type MaxCurrencyName: Get<u32>;
+
+		/// Maximum number of members in a pool
+		#[pallet::constant]
+		type MaxMembersPerPool: Get<u32>;
+
+		/// Maximum number of selectable targets for a pool
+		#[pallet::constant]
+		type MaxTargets: Get<u32>;
+
+		/// Maximum number of parallel partial unbonds
+		#[pallet::constant]
+		type MaxUnbonding: Get<u32>;
+
+		/// Maximum of number of concurrent pending slashes for a pool
+		#[pallet::constant]
+		type MaxSlashesPerPool: Get<u32>;
+
+		/// Maximum number of pools behind a validator, mostly used to set bounds
+		#[pallet::constant]
+		type MaxPoolsPerValidator: Get<u32>;
+
+		/// Period for funds to be available after unbonding
+		#[pallet::constant]
+		type BondingDuration: Get<EraIndex>;
+
+		/// Number of era for which to keep Fusion data
+		#[pallet::constant]
+		type HistoryDepth: Get<u32>;
+
+		/// A provider that gives the information from the staking pallet.
+		type StakingFusionDataProvider: StakingFusionDataProvider<Self::AccountId>;
+	}
+
+	#[pallet::pallet]
+	pub struct Pallet<T>(_);
+
+	/// Stores the total value locked in avail and the maximum total value locked authorized
+	#[pallet::storage]
+	#[pallet::getter(fn tvl_data)]
+	pub type TotalValueLockedData<T> = StorageValue<_, TVLData<T>, ValueQuery>;
+
+	/// Stores all the fusion currencies
+	#[pallet::storage]
+	#[pallet::getter(fn fusion_currencies)]
+	pub type FusionCurrencies<T: Config> =
+		StorageMap<_, Twox64Concat, CurrencyId, FusionCurrency<T>, OptionQuery>;
+
+	/// Stores all the fusion pools
+	#[pallet::storage]
+	#[pallet::getter(fn fusion_pools)]
+	pub type FusionPools<T: Config> =
+		StorageMap<_, Twox64Concat, PoolId, FusionPool<T>, OptionQuery>;
+
+	/// Mapping from the pools funds account address to the pool id
+	#[pallet::storage]
+	#[pallet::getter(fn fusion_pool_account_to_id)]
+	pub type FusionPoolsAccountToId<T: Config> =
```

Comment:
Maybe we can avoid `Fusion` prefixes for most storages when you do cleaning next time?